### PR TITLE
NMS-14966: Add smoke test for dual timeseries writes

### PR DESCRIPTION
**Pull request by amay94**
The `inmemory-timeseries-plugin` feature is a repackaged version of this: https://github.com/OpenNMS/opennms-integration-api/tree/master/test-suites/tss-tests. I pulled it into the project so I could install it as a feature for testing.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14966

### CODE REVIEW

**Review by cgorantla — DISMISSED** [review]
LGTM!

**Review by christianpape — CHANGES_REQUESTED** [review]

**Review by christianpape — DISMISSED** [review]
LGTM.

File: features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            throw new IllegalArgumentException("We expect at least one TagMatcher but none was given.");
+        }
+        return data.keySet()
+                .stream()
+                .filter(metric -> this.matches(tagMatchers, metric))
+                .collect(Collectors.toList());
+    }
+
+    /** Each matcher must be matched by at least one tag. */
+    private boolean matches(final Collection<TagMatcher> matchers, final Metric metric) {
```
Comment (fooker) [inline features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java:95]: can be static?

File: features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+        for(TagMatcher matcher : matchers) {
+            if(searchableTags.stream().noneMatch(t -> this.matches(matcher, t))) {
+                return false; // this TagMatcher didn't find any matching tag => this Metric is not part of search result;
+            }
+        }
+        return true; // all matched
+    }
+
+    private boolean matches(final TagMatcher matcher, final Tag tag) {
```
Comment (fooker) [inline features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java:107]: can be static?

File: features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+public class InMemoryStorage implements TimeSeriesStorage {
+
+    private final Map<Metric, Collection<DataPoint>> data;
+
+    public InMemoryStorage () {
+        data = new ConcurrentHashMap<>();
+    }
+
+    public final Map<Metric, Collection<DataPoint>> getAllMetrics() {
+        return data;
```
Comment (fooker) [inline features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java:70]: Wrap with `Collections.immutableMap`

File: features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+/**
+ * Simulates a TimeSeriesStorage in memory (ConcurrentHashMap). The implementation is super simple and not very efficient.
+ * For testing and evaluating purposes only, not for production.
+ */
+public class InMemoryStorage implements TimeSeriesStorage {
+
+    private final Map<Metric, Collection<DataPoint>> data;
+
+    public InMemoryStorage () {
+        data = new ConcurrentHashMap<>();
```
Comment (fooker) [inline features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java:66]: Assign in field decl and drop constructor.

File: features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+    @Override
+    public String toString() {
+        return this.getClass().getName();
+    }
+
+    /** Exposes the internal data, for validation in tests. */
+    public Map<Metric, Collection<DataPoint>> getData() {
+        return data;
+    }
```
Comment (fooker) [inline features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/InMemoryStorage.java:173]: Isn't that the save as `getAllMetrics`?

File: features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/shell/GetMetricsCommand.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    private TimeSeriesStorage storage;
+
+    @Override
+    public Object execute() {
+        ((InMemoryStorage)storage).getAllMetrics().entrySet().forEach(entry -> {
+            System.out.println(String.format("Metric <%s> has %d data points",
+                    entry.getKey().getKey(),
+                    entry.getValue().size()));
+            entry.getValue().stream().forEach(dataPoint ->
+                System.out.println(String.format("\t%d:%f", dataPoint.getTime().toEpochMilli(), dataPoint.getValue())));
```
Comment (fooker) [inline features/inmemory-timeseries-plugin/src/main/java/org/opennms/features/timeseries/plugin/shell/GetMetricsCommand.java:57]: `System.out.printf` insted?

**Review by fooker — CHANGES_REQUESTED** [review]
Just a bunch of minor code-style nits.

**Review by RangerRick — APPROVED** [review]
I was going to suggest making a new `features-test.xml` for this rather than sticking the feature in `features.xml` but honestly, it's a lot of work for just a semantic change with no real difference. :D

LGTM

**Pull request merged**